rpc/rpc-transport/socket/src/socket.c

```diff
@@ -982,6 +984,11 @@ __socket_server_bind(rpc_transport_t *this)
                    this->myinfo.identifier, strerror(errno));
             if (errno == EADDRINUSE) {
                 gf_log(this->name, GF_LOG_ERROR, "Port is already in use");
+                retries--;
+                if (retries) {
+                    sleep(1);
+                    goto retry;
```

Comment:
A retry doesn't guarantee that the port will be free after 3 seconds (most probably it won't be).
Glusterd has a check for EADDRINUSE error and restarts the brick on a different port if the current port is occupied.
This [patch](https://github.com/gluster/glusterfs/pull/2090) did fix that, however, the else part was missed out in it.
So, I think a better fix would be to save the errno value in the `ret` variable as done on line: 941. That should fix the issue.

Reply:
> A retry doesn't guarantee that the port will be free after 3 seconds (most probably it won't be).
> Glusterd has a check for EADDRINUSE error and restarts the brick on a different port if the current port is occupied.
> This [patch](https://github.com/gluster/glusterfs/pull/2090) did fix that, however, the else part was missed out in it.
> So, I think a better fix would be to save the errno value in the `ret` variable as done on line: 941. That should fix the issue.

@nik-redhat 
That may be not applicable to AF_UNIX socket. Let me take a closer look at the code again.

Reply:
@nik-redhat 
The failed path is for AF_UNIX. Firstly unlink/close it，before reuse the socket.
EADDRINUSE is a special errno for rechoice port in glusterd. Return it when no needed，may be puzzled.

Reply:
@xhernandez 
When a file (e.g. socket) is closed, the work is handed over to the workqueue for asynchronous execution by calling fput in linux kernel.
If the workqueue thread is busy, the delay may be more. In this case, bind will fail when the socket is not actually released. Retry can deal with this situation.

Reply:
I'm not an expert on the kernel side, but from POSIX as soon as the `unlink()` request completes and returns to user code, the file shouldn't be there anymore and a new file with the same name should be able to be created without issues, even if the kernel still has some references to clean from the old file. So creating a new file with the same name shouldn't cause any issue, right ? this means that bind should succeed. Am I missing something ?

I'm just trying to understand the situation. The patch seems fine, but I would like to understand when it's needed, since the same thing could happen in other places.

Reply:
@xhernandez 
For AF_UNIX socket, ulink the files does not affect the use of connected socket. Before the socket is actually released, the socket state can still be observed using ss，and another bind will failed with "already in use".
If there is a large backlog in the same workqueue, some work (e.g. socket release) will be delayed. I have been encountered XFS affecting the release of socket.

Reply:
Thanks for the explanation.